include/oneapi/dpl/internal/dynamic_selection_impl/round_robin_policy_impl.h

```diff
@@ -0,0 +1,128 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ROUND_ROBIN_POLICY_IMPL_H
+#define _ROUND_ROBIN_POLICY_IMPL_H
+#pragma once
+
+#include <atomic>
+#include <ostream>
+#include "oneapi/dpl/internal/dynamic_selection_impl/scoring_policy_defs.h"
+
+namespace oneapi {
+namespace dpl{
+namespace experimental{
+
+  template <typename Scheduler>
+  struct round_robin_policy_impl {
+    using scheduler_t = Scheduler;
+    using native_resource_t = typename scheduler_t::native_resource_t;
+    using execution_resource_t = typename scheduler_t::execution_resource_t;
+    using native_sync_t = typename scheduler_t::native_sync_t;
+    using universe_container_t = typename scheduler_t::universe_container_t;
+    using selection_handle_t = oneapi::dpl::experimental::nop_selection_handle_t<execution_resource_t>;
+    using universe_container_size_t = typename universe_container_t::size_type;
+
+    std::shared_ptr<scheduler_t> sched_;
+    universe_container_t universe_;
+    universe_container_size_t num_contexts_;
+    std::atomic<universe_container_size_t> next_context_;
+
+    round_robin_policy_impl() : sched_{std::make_shared<scheduler_t>()}  {
+      universe_ = oneapi::dpl::experimental::property::query(*sched_, property::universe);
+      num_contexts_ = universe_.size();
+      next_context_ = 0;
+    }
+
+    round_robin_policy_impl(universe_container_t u) : sched_{std::make_shared<scheduler_t>()}  {
+      oneapi::dpl::experimental::property::report(*sched_, property::universe, u);
+      universe_ = oneapi::dpl::experimental::property::query(*sched_, property::universe);
```

Comment:
Could class fields be initialized in member-init-list? Do the order that `report` should be done before fields initialization matter?

Reply:
I believe the justification for this is that the "report" is a request for the scheduler to change its universe. But it may not want to or may not be able to. And so even after reporting/requesting, it needs to query what the universe actually is.